src/autoscaling/snapshotter.js

```diff
@@ -0,0 +1,267 @@
+            else break;
+        }
+        return sample;
+    }
+
+    /**
+     * Creates a snapshot of current memory usage.
+     * @param {Function} intervalCallback
+     * @return {Promise}
+     * @ignore
+     */
+    async _snapshotMemory(intervalCallback) {
+        try {
+            const now = new Date();
+            const memInfo = await getMemoryInfo();
+            const { mainProcessBytes, childProcessesBytes } = memInfo;
+            this._pruneSnapshots(this.memorySnapshots, now);
+
+            const usedBytes = mainProcessBytes + childProcessesBytes;
+            const snapshot = Object.assign(memInfo, {
+                createdAt: now,
+                isOverloaded: usedBytes / this.maxMemoryBytes > this.maxUsedMemoryRatio,
+            });
+
+            this.memorySnapshots.push(snapshot);
+        } catch (err) {
+            log.exception(err, 'Snapshotter: Memory snapshot failed.');
+        } finally {
+            intervalCallback();
+        }
+    }
+
+    /**
+     * Creates a snapshot of current event loop delay.
+     * @param {Function} intervalCallback
+     * @return {Promise}
+     * @ignore
+     */
+    _snapshotEventLoop(intervalCallback) {
+        try {
+            const now = new Date();
+            this._pruneSnapshots(this.eventLoopSnapshots, now);
+            const { createdAt } = this.eventLoopSnapshots[this.eventLoopSnapshots.length - 1];
+            const delta = now - createdAt - this.eventLoopSnapshotIntervalMillis;
+
+            const snapshot = {
+                createdAt: now,
+                isOverloaded: false,
+                exceededMillis: Math.max(delta - this.maxBlockedMillis, 0),
+            };
+
+            if (delta > this.maxBlockedMillis) snapshot.isOverloaded = true;
+            this.eventLoopSnapshots.push(snapshot);
+        } catch (err) {
+            log.exception(err, 'Snapshotter: Event Loop snapshot failed.');
+        } finally {
+            intervalCallback();
+        }
+    }
+
+    /**
+     * Creates a snapshot of current CPU usage.
+     * @param {Object} cpuInfoEvent
+     * @return {Promise}
+     * @ignore
+     */
+    _snapshotCpu(cpuInfoEvent) {
+        const createdAt = (new Date(cpuInfoEvent.createdAt));
+        this._pruneSnapshots(this.cpuSnapshots, createdAt);
+        this.cpuSnapshots.push({
+            createdAt,
+            isOverloaded: cpuInfoEvent.isCpuOverloaded,
+        });
+    }
+
+    /**
+     * Removes snapshots that are older than the snapshotHistorySecs option
+     * from the array (destructively - in place).
+     * @param {Array} snapshots
+     * @param {Date} now
+     * @ignore
+     */
+    _pruneSnapshots(snapshots, now) {
+        let oldCount = 0;
+        for (let i = 0; i < snapshots.length; i++) {
+            const { createdAt } = snapshots[i];
+            if (now - createdAt > this.snapshotHistoryMillis) oldCount++;
+            else break;
+        }
+        snapshots.splice(0, oldCount);
```

Comment:
This is O(n) operation. For this kind of structure it's better to use linked-list (we have one in apify-shared), and then just remove elements from the start. It will be O(1) always

Reply:
I researched this and one of Node's creators actually says that native Array is represented differently in the compiled code, based on what operations are used on it. If you use it as a queue (push, shift), it will automatically optimize it. https://github.com/isaacs/fast-list

Reply:
But will this work with splice? We don't know, so perhaps it's better to just use LinkedList, to make sure it's fine